scipy/optimize/minpack.py

```diff
@@ -701,29 +701,25 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         raise ValueError("Method 'lm' only works for unconstrained problems. "
                          "Use 'trf' or 'dogbox' instead.")
 
+    # optimization may produce garbage for float32 inputs, cast them to float64
+
     # NaNs can not be handled
     if check_finite:
-        ydata = np.asarray_chkfinite(ydata)
+        ydata = np.asarray_chkfinite(ydata, float)
     else:
-        ydata = np.asarray(ydata)
+        ydata = np.asarray(ydata, float)
 
     if isinstance(xdata, (list, tuple, np.ndarray)):
         # `xdata` is passed straight to the user-defined `f`, so allow
         # non-array_like `xdata`.
         if check_finite:
-            xdata = np.asarray_chkfinite(xdata)
+            xdata = np.asarray_chkfinite(xdata, float)
```

Comment:
My comment doesn't belong here, but still...

The docstring for `xdata` needs to be updated.

Reply:
done

Reply:
I think it needs to be explicit.

`xdata : array_like or None`

Reply:
It can really be any object, so I edited the docstring accordingly.